### ISSUE TITLE
Views block pager issue

### ISSUE THREAD

**Issue by pyrello** [uiowa/uiowa#3115]
# Problem
As identified in #2730, there is a problem when passing block styles to the view and clicking on a pager link in a views block. On the second page of results, any styles that were passed through to the view are lost. I'm putting this separate issue together to make it easier to see the relevant information to try to resolve this.

# Steps to reproduce
The easiest way to see this issue is to test #2730 locally.

# Proposed solution
The best ideas we have so far about why this is happening are:
1. We are failing to invalidate a cache or properly set a cache when adding LB styles to the view.
2. The method we are using to set the styles - using a custom `$view->display_hander->setOption()` only gets called when the block is rendered as part of the normal page rendering process. When called via AJAX the section in question doesn't get triggered.

Some research has uncovered this issue: https://www.drupal.org/project/drupal/issues/2605218, which has a couple patches that have been tried on #2730 to no avail.

**Comment by joewhitsitt** [comment]
Wonder if we could store these classes on the view another way and have that called since the display_handler stuff doesn't seem to? hidden header or footer?

**Comment by pyrello** [comment]
In testing this a bit more, part of the problem seems to be that `HOOK_preprocess_block()` never gets called when the view is loaded via AJAX. So, it follows that our processing there is just not populating the correct classes in the view for rendering.

**Comment by pyrello** [comment]
I also tested this by adding the following functions and adding breakpoints to them to see if the AJAX view triggered either of them to be fired: `sitenow_people_block_view_alter` and `sitenow_people_block_content_view_alter`. Neither one seems to get triggered as part of the AJAX view generation.

**Comment by pyrello** [comment]
Okay... I think I may have finally wrapped my brain around why this is an issue:

In a normal page load scenario, as we build the page, we build each block. Each block that contains a view in turn triggers that view to be rendered. In that process, information can be passed between the block instance and the view. When paging occurs as an AJAX request, what is being requested is the updated View with altered parameters for displaying whatever set of results would be displayed on the page being requested. The block is not generated because it is not being replaced via AJAX, just the view. In the controller that updates the View, you can see that the element ID that it is targeting for replacement is inside the block:

https://git.drupalcode.org/project/drupal/-/blob/9.2.x/core/modules/views/src/Controller/ViewAjaxController.php#L202

Here is a fragment of the View block output where you can see the targeted class in question:
![Screen Shot 2021-02-20 at 1 29 37 PM](https://user-images.githubusercontent.com/709009/108606481-e716bd00-737f-11eb-921f-bca10bcf7b5c.png)

The patch at https://www.drupal.org/project/drupal/issues/2605218 causes the `preBlockBuild()` method to be triggered on AJAX requests and in so doing causes the block instance to be instantiated. But it doesn't make the block available to be accessed through any mechanism (event or hook) that I can find.

I think for this to work, we would need a core patch that adds an event that fires during `preBlockBuild()` so that we can subscribe to it and handle passing the data from the block to the view there. Views doesn't currently appear to provide any events at all, so this seems to be new territory.

**Comment by joewhitsitt** [comment]
You got further than me. I was investigating actually making a proper views plugin for heading_size and lb_parent_styles as well as trying to stick the classes to a views core bucket like header area or css_class. While I did learn how to make a DisplayExtenderPlugin none of my attempts stuck to a pager ajax operation.

Turning off AJAX for these works but all of the pagers on the page advance as well. If we could figure out a way to assign a unique pager id to the view and have the pagers act independently, maybe this would be a path forward for either implementation?

**Comment by pyrello** [comment]
Even if you were able to set it up so that the settings came from the view, the settings that would be in play would be the block instance configuration, so that editors could set different settings per block. And we would run into the same issues.

**Comment by pyrello** [comment]
I took a stab at creating the core patch that I think would be a requisite first step to being able to make this work to pass data from block to view: https://git.drupalcode.org/issue/drupal-3200029/-/compare/9.2.x...3200029-views-block-display-prebuild-event. That combined with the patch above, along with some additional custom code in a PreBlockBuild event subscriber to get the block information and pass it to the views is a plausible path forward.

**Comment by pyrello** [comment]
The previous link I posted leads to a hot mess (tl;dr I got mixed up between `drupal/drupal` and `drupal/core`). I finally got a properly formatted patch attached to the issue: https://www.drupal.org/project/drupal/issues/3200029#comment-14011712

**Comment by pyrello** [comment]
I have figured out a path forward on this in #2702 that I think is fairly robust, which depends on the patch in https://www.drupal.org/project/drupal/issues/2605218. That patch allows the `preBuildBlock` method to run even when the views is built via AJAX. The headline field being part of the views block means that it is fairly trivial to pass the `child_heading_size` setting through. The Layout Builder styles are a bit trickier, but the idea is the same. We tap into the block submit function and save the Layout Builder styles as configuration on the block - which is not what currently happens - and then those styles become available during the `preBlockBuild` method.

I'm still working on the best way to pass those through to the individual view results, but I am not confident that it can be done because I have been able to access them on the `$variables['view']` object in `HOOK_preprocess_views_view`, which runs every time a view is generated.
